File: ios/LibraryImageResize.m
//
//  LibraryImageResize.m
//  MediaLibrary
//

#import "LibraryImageResize.h"
#import "LibraryImageSize.h"
#import <React/RCTImageUtils.h>

@implementation LibraryImageResize

+ (nullable NSString *)resizeWithUri:(NSString *)uri
                               width:(NSNumber *)width
                              height:(NSNumber *)height
                              format:(NSString *)format
                      resultSavePath:(NSString *)resultSavePath {
    UIImage *image = [LibraryImageSize imageWithPath:uri];
    if (!image) {
        return @"LibraryImageResize.image.notExists";
    }

    CGFloat imageWidth = image.size.width;
    CGFloat imageHeight = image.size.height;
    CGFloat imageRatio = imageWidth / imageHeight;
    CGSize targetSize = CGSizeZero;

    if ([width floatValue] >= 0) {
        targetSize.width = [width floatValue];
        targetSize.height = targetSize.width / imageRatio;
    }

    if ([height floatValue] >= 0) {
        targetSize.height = [height floatValue];
        if (targetSize.width <= 0) {
            targetSize.width = imageRatio * targetSize.height;
        }
    }

    UIGraphicsBeginImageContextWithOptions(targetSize, NO, 1.0);
    [image drawInRect:CGRectMake(0, 0, targetSize.width, targetSize.height)];
    UIImage *finalImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();

    if (!finalImage) {
        return @"LibraryImageResize.image.emptyContext";
    }

    return [LibraryImageSize saveImage:finalImage format:format path:resultSavePath];
}

+ (nullable NSString *)cropWithUri:(NSString *)uri
                                 x:(NSNumber *)x
                                 y:(NSNumber *)y
                             width:(NSNumber *)width
                            height:(NSNumber *)height
                            format:(NSString *)format
                    resultSavePath:(NSString *)resultSavePath {
    UIImage *image = [LibraryImageSize imageWithPath:uri];
    if (!image) {
        return @"LibraryImageResize.crop.notExists";
    }

    CGSize originalSize = image.size;
    CGFloat fX = [x floatValue] * originalSize.width;
    CGFloat fY = [y floatValue] * originalSize.height;
    CGFloat fWidth = [width floatValue];
    CGFloat fHeight = [height floatValue];

    if (fX + fWidth > image.size.width) {
        fX = image.size.width - fWidth;
    }

    if (fY + fHeight > image.size.height) {
        fY = image.size.height - fHeight;
    }

    CGSize targetSize = CGSizeMake(fWidth, fHeight);
    CGRect targetRect = CGRectMake(-fX, -fY, image.size.width, image.size.height);
    CGAffineTransform transform = RCTTransformFromTargetRect(image.size, targetRect);
    UIImage *finalImage = RCTTransformImage(image, targetSize, image.scale, transform);

    if (!finalImage) {
        return @"LibraryImageResize.crop.errorTransform";
    }

    return [LibraryImageSize saveImage:finalImage format:format path:resultSavePath];
}

@end

File: ios/LibraryImageResize.swift
//
//  LibraryImageResize.swift
//  MediaLibrary
//
//  Created by sergeymild on 05/08/2023.
//  Copyright © 2023 Facebook. All rights reserved.
//

import Foundation
import React

@objc
public class LibraryImageResize: NSObject {
    @objc
    public static func resize(
        uri: NSString,
        width: NSNumber,
        height: NSNumber,
        format: NSString,
        resultSavePath: NSString
    ) -> String? {
        guard let image = LibraryImageSize.image(path: uri as String) else {
            return "LibraryImageResize.image.notExists"
        }
        let imageWidth = image.size.width
        let imageHeight = image.size.height
        let imageRatio = imageWidth / imageHeight
        var targetSize: CGSize = .zero
        
        if (width.floatValue >= 0) {
            targetSize.width = CGFloat(width.floatValue)
            targetSize.height = targetSize.width / imageRatio
        }
        
        if (height.floatValue >= 0) {
            targetSize.height = CGFloat(height.floatValue)
            targetSize.width = targetSize.width <= 0 ? imageRatio * targetSize.height : targetSize.width
        }
        UIGraphicsBeginImageContextWithOptions(targetSize, false, 1.0)
        image.draw(in: .init(origin: .zero, size: targetSize))
        let finalImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        guard let finalImage = finalImage else {
            return "LibraryImageResize.image.emptyContext"
        }
        
        return LibraryImageSize.save(image: finalImage, format: format, path: resultSavePath)
    }
    
    @objc
    public static func crop(
        uri: NSString,
        x: NSNumber,
        y: NSNumber,
        width: NSNumber,
        height: NSNumber,
        format: NSString,
        resultSavePath: NSString
    ) -> String? {
        guard let image = LibraryImageSize.image(path: uri as String) else {
            return "LibraryImageResize.crop.notExists"
        }
        let originalSize = image.size
        var fX = CGFloat(x.floatValue) * originalSize.width
        var fY = CGFloat(y.floatValue) * originalSize.height
        let fWidth = CGFloat(width.floatValue)
        let fHeight = CGFloat(height.floatValue)
        
        if (fX + fWidth > image.size.width) {
            fX = image.size.width - fWidth;
        }

        if (fY + fHeight > image.size.height) {
            fY = image.size.height - fHeight;
        }
        
        let targetSize = CGSize(width: fWidth, height: fHeight)
        let targetRect = CGRect(x: -fX, y: -fY, width: image.size.width, height: image.size.height)
        let transform = RCTTransformFromTargetRect(image.size, targetRect)
        guard let finalImage = RCTTransformImage(image, targetSize, image.scale, transform) else {
            return "LibraryImageResize.crop.errorTransform"
        }
        return LibraryImageSize.save(image: finalImage, format: format, path: resultSavePath)
    }
}
